### DISCUSSION TITLE
Python

### DISCUSSION THREAD

**Discussion by collinanderson** [jart/cosmopolitan#348] [General]
Hi All,

The cosmopolitan project is amazing. I've been following it off and on for about a year and it's incredible to watch. Thanks for all of your work on it.

As a Python web developer I'm curious: why did you use Python 3.6/3.7 as a starting point - why not something newer?

Thanks,
Collin

P.S. Any plans for an alpine-equivalent distribution? :) Or at least a package-manager + repository? :)

**Comment by collinanderson** [comment]
Ohh I finally found what I think was the answer: "Python 3.7 and above cannot be built without threads." https://github.com/jart/cosmopolitan/issues/141#issuecomment-894710007

"Python 3.7 and above require pthread by default, which Cosmopolitan does not currently support. So I picked Python3.6. I would like to use the newest version also, but I do not know of a way to port them yet."
https://github.com/jart/cosmopolitan/issues/141#issuecomment-905043548

Makes sense. Thanks!

**Comment by ahgamut** [comment]
@collinanderson I looked at the source code of Python 3.9.7 today, and it is possible to build it by disabling pthread everywhere:

https://github.com/ahgamut/cpython/tree/cosmo_py39 (clone the repo, get the latest build of Cosmopolitan Libc in the `libcosmo` folder, run superconfigure, run `python.com`)

The above repo builds a APE Python 3.9 which works on linux. It's not as polished at APE Python 3.6, still in an experimental stage (`_socket` isn't there, I need to test how lack of pthreads affects the `_thread` module, none of the Cosmopolitan Libc goodies by @jart are available), but it works, and can likely get there:

![2022-02-05_17-15-31_1366x736](https://user-images.githubusercontent.com/41098605/152641038-8b58d9c0-8690-466d-a242-c83c892bc7cc.png)

**Reply by collinanderson** [reply]
Sweet. That's exciting. Thanks.

**Comment by ahgamut** [comment]
@collinanderson 

Now that @jart has implemented pthreads in Cosmopolitan Libc, we can have Python versions with multithreading!

Python 3.11.0rc1 was released on August 5. Here's an experimental port of Python 3.11.0rc1 to Cosmopolitan Libc: https://github.com/ahgamut/cpython/tree/cosmo_py311

**Reply by collinanderson** [reply]
Sweet!
